File: kotlin-sample-app/src/main/java/org/remdev/android/kotlinapp/models/InteractionResult.java
package org.remdev.android.kotlinapp.models;

import org.remdev.android.kotlinapp.annotations.POJO;
import org.remdev.android.kotlinapp.constants.StatusCodes;
import org.remdev.android.kotlinapp.rest.ToolboxKt;
import org.remdev.timlog.Log;
import org.remdev.timlog.LogFactory;

import java.util.HashMap;
import java.util.Map;


public class InteractionResult<T> implements JsonSerializable<InteractionResult<T>> {

    private static final Log log = LogFactory.create(InteractionResult.class);

    private static final String FIELD_CODE = "code";
    private static final String FIELD_ERROR_CODE = "errorCode";
    private static final String FIELD_PAYLOAD_CLASS = "payloadClass";
    private static final String FIELD_PAYLOAD = "payload";

    private int code;
    private int errorCode;
    @POJO
    private T payload;

    public InteractionResult() { }

    public int getCode() {
        return code;
    }

    public void setCode(int code) {
        this.code = code;
    }

    public int getErrorCode() {
        return errorCode;
    }

    public void setErrorCode(int errorCode) {
        this.errorCode = errorCode;
    }

    public T getPayload() {
        return payload;
    }

    public void setPayload(T payload) {
        this.payload = payload;
    }

    public boolean isSuccessful() {
        return code == StatusCodes.CODE_OK;
    }

    @Override
    public String toJson() {
        Map<String, String> map = new HashMap<>();
        map.put(FIELD_CODE, String.valueOf(code));
        map.put(FIELD_ERROR_CODE, String.valueOf(errorCode));
        if (payload != null) {
            map.put(FIELD_PAYLOAD_CLASS, payload.getClass().getName());
            map.put(FIELD_PAYLOAD, ToolboxKt.toJson(payload));
        }
        return ToolboxKt.toJson(map);
    }

    @Override
    @SuppressWarnings("unchecked")
    public InteractionResult<T> fromJson(String json) {
        return parseFromJson(json);
    }

    public <R> InteractionResult<R> changePayload(R newPayload) {
        InteractionResult<R> result = new InteractionResult<>();
        result.setCode(this.getCode());
        result.setErrorCode(this.getErrorCode());
        result.setPayload(newPayload);
        return result;
    }

    public static InteractionResult success() {
        return success(null);
    }

    public static <T> InteractionResult<T> success(@POJO T payload) {
        InteractionResult<T> result = new InteractionResult<>();
        result.code = StatusCodes.CODE_OK;
        result.errorCode = 0;
        result.setPayload(payload);
        return result;
    }

    public static <T> InteractionResult<T> error() {
        return error(0);
    }

    public static <T> InteractionResult<T> error(int errorCode) {
        return error(errorCode, null);
    }

    public static <T> InteractionResult<T> error(@POJO T payload) {
        return error(0, payload);
    }

    public static <T> InteractionResult<T> error(int errorCode, @POJO T payload) {
        InteractionResult<T> result = new InteractionResult<>();
        result.code = StatusCodes.CODE_ERROR;
        result.errorCode = errorCode;
        result.setPayload(payload);
        return result;
    }

    @SuppressWarnings("unchecked")
    public static<T> InteractionResult<T> parseFromJson(String json) {
        if (json == null) {
            return null;
        }
        Map map = ToolboxKt.fromJson(json, Map.class);
        String code = String.valueOf(map.get(FIELD_CODE));
        String errorCode = String.valueOf(map.get(FIELD_ERROR_CODE));
        String className = (String) map.get(FIELD_PAYLOAD_CLASS);
        String payloadJson = (String) map.get(FIELD_PAYLOAD);
        if (code == null || errorCode == null) {
            throw new IllegalArgumentException("Invalid json: " + json);
        }
        int resCode = Integer.parseInt(code);
        int resErrorCode = Integer.parseInt(errorCode);
        T payload = null;
        if (className != null) {
            try {
                Class<T> cls = (Class<T>) Class.forName(className);
                payload = ToolboxKt.fromJson(payloadJson, cls);
            } catch (ClassNotFoundException e) {
                log.e("Could not found class %s", className);
                throw new IllegalArgumentException("Invalid json: "+ json);
            }
        }
        InteractionResult<T> res = new InteractionResult<>();
        res.setCode(resCode);
        res.setErrorCode(resErrorCode);
        res.setPayload(payload);
        return res;
    }
}

File: commander/src/main/java/org/remdev/android/commander/models/InteractionResult.kt
package org.remdev.android.commander.models;

import org.remdev.android.commander.ResultCodes
import org.remdev.android.commander.annotations.POJO
import org.remdev.android.commander.utils.ConversionUtils
import org.remdev.timlog.Log
import org.remdev.timlog.LogFactory

open class InteractionResult<T: Any?> : JsonSerializable<InteractionResult<T>> {

    var code = 0
    var errorCode = 0
    @POJO
    var payload : T? = null

    override fun toJson(): String {
        val map : MutableMap<String, String> = mutableMapOf(
                FIELD_CODE to code.toString(),
                FIELD_ERROR_CODE to errorCode.toString()
        )
        payload?.let {
            map.put(FIELD_PAYLOAD_CLASS, ConversionUtils.getClass(it!!).name)
            map.put(FIELD_PAYLOAD, ConversionUtils.toJson(it))
        }
        return ConversionUtils.toJson(map)
    }

    companion object {
        val log : Log = LogFactory.create(InteractionResult::class.java)
        const val FIELD_CODE = "code"
        const val FIELD_ERROR_CODE = "errorCode"
        const val FIELD_PAYLOAD_CLASS = "payloadClass"
        const val FIELD_PAYLOAD = "payload"

        fun success() : InteractionResult<Any?> = success(null)

        fun <R> success(payload: R) : InteractionResult<R> {
            val res = InteractionResult<R>()
            res.code = ResultCodes.CODE_OK
            res.errorCode = 0
            res.payload = null
            return res
        }

        fun error() : InteractionResult<Any?> = error(0)

        fun error(errorCode: Int): InteractionResult<Any?> = error(errorCode, null)

        fun <R> error(@POJO payload: R?): InteractionResult<Any?> = error(0, payload)

        fun <R> error(errorCode: Int, @POJO payload: R?): InteractionResult<R> {
            val result = InteractionResult<R>()
            result.code = ResultCodes.CODE_ERROR
            result.errorCode = errorCode
            result.payload = payload
            return result
        }

        fun <T> parseFromJson(json: String?) : InteractionResult<T>? {
            if (json == null) {
                return null
            }
            val map = ConversionUtils.fromJson(json, Map::class.java)
            if (map == null) {
                return null;
            }
            val code = (map[FIELD_CODE] as String?)?.toInt()
            val errorCode = (map[FIELD_ERROR_CODE] as String?)?.toInt()
            val className = map[FIELD_PAYLOAD_CLASS] as String?
            val payloadJson = map[FIELD_PAYLOAD] as String?
            if (code == null || errorCode == null) {
                throw IllegalArgumentException("Invalid json: " + json);
            }
            var payload : T? = null
            if (className != null) {
                try {
                    @Suppress("UNCHECKED_CAST")
                    val clazz = Class.forName(className) as Class<T>
                    payload = ConversionUtils.fromJson(payloadJson, clazz)
                } catch (e: ClassNotFoundException) {
                    log.e("Could not found class %s", className)
                    throw IllegalArgumentException("Invalid json: "+ json)
                }
            }
            val res = InteractionResult<T>();
            res.code = code
            res.errorCode = errorCode
            res.payload = payload
            return res;
        }
    }
}